server/src/main/java/io/druid/segment/indexing/DataSchema.java

```diff
@@ -39,33 +44,72 @@
-  private final InputRowParser parser;
+  private final Map<String, Object> parser;
   private final AggregatorFactory[] aggregators;
   private final GranularitySpec granularitySpec;
 
+  private final ObjectMapper jsonMapper;
+
   @JsonCreator
   public DataSchema(
       @JsonProperty("dataSource") String dataSource,
-      @JsonProperty("parser") InputRowParser parser,
+      @JsonProperty("parser") Map<String, Object> parser,
       @JsonProperty("metricsSpec") AggregatorFactory[] aggregators,
-      @JsonProperty("granularitySpec") GranularitySpec granularitySpec
+      @JsonProperty("granularitySpec") GranularitySpec granularitySpec,
+      @JacksonInject ObjectMapper jsonMapper
   )
   {
-    Preconditions.checkNotNull(dataSource, "dataSource cannot be null. Please provide a dataSource.");
+    this.jsonMapper = Preconditions.checkNotNull(jsonMapper, "null ObjectMapper.");
+    this.dataSource = Preconditions.checkNotNull(dataSource, "dataSource cannot be null. Please provide a dataSource.");
+    this.parser = parser;
 
-    this.dataSource = dataSource;
+    if (aggregators.length == 0) {
+      log.warn("No metricsSpec has been specified. Are you sure this is what you want?");
+    }
+    this.aggregators = aggregators;
+
+    if (granularitySpec == null) {
+      log.warn("No granularitySpec has been specified. Using UniformGranularitySpec as default.");
+      this.granularitySpec = new UniformGranularitySpec(null, null, null);
+    } else {
+      this.granularitySpec = granularitySpec;
+    }
+  }
 
-    final Set<String> dimensionExclusions = Sets.newHashSet();
+  @JsonProperty
+  public String getDataSource()
+  {
+    return dataSource;
+  }
+
+  @JsonProperty("parser")
+  public Map<String, Object> getParserMap()
+  {
+    return parser;
+  }
+
+  @JsonIgnore
+  public InputRowParser getParser()
+  {
+    if(parser == null) {
+      log.warn("No parser has been specified");
+      return null;
+    }
+
+    InputRowParser inputRowParser = jsonMapper.convertValue(this.parser, InputRowParser.class);
+
+    Set<String> dimensionExclusions = Sets.newHashSet();
     for (AggregatorFactory aggregator : aggregators) {
       dimensionExclusions.addAll(aggregator.requiredFields());
     }
-    if (parser != null && parser.getParseSpec() != null) {
-      final DimensionsSpec dimensionsSpec = parser.getParseSpec().getDimensionsSpec();
-      final TimestampSpec timestampSpec = parser.getParseSpec().getTimestampSpec();
+
+    if (inputRowParser.getParseSpec() != null) {
+      DimensionsSpec dimensionsSpec = inputRowParser.getParseSpec().getDimensionsSpec();
+      TimestampSpec timestampSpec = inputRowParser.getParseSpec().getTimestampSpec();
 
       // exclude timestamp from dimensions by default, unless explicitly included in the list of dimensions
       if (timestampSpec != null) {
-        final String timestampColumn = timestampSpec.getTimestampColumn();
+        String timestampColumn = timestampSpec.getTimestampColumn();
```

Comment:
sounds reasonable, will add the finals back considering this is copy/pasted code and shouldn't be changed for no reason.